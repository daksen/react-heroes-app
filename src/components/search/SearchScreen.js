import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useHistory, useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../heroes/HeroCard';

export const SearchScreen = () => {

  const history = useHistory();
  const location = useLocation();
  const { q = '' } = queryString.parse(location.search);
  const [{search}, handleInputChange] = useForm({
    'search': q
  });

  const heroesFiltered = useMemo(() => getHeroesByName(q), [q])

  const hanldeSearch = (e) => {
    e.preventDefault();
    history.push(`?q=${search}`);
  }

  return (
    <div className="container mt-5">
      <div className="row">
        <div className="col-12 col-md-5 mb-4 mb-md-0">
          <h4>Search Form</h4>
          <hr/>
          <form onSubmit={ hanldeSearch } className="animate__animated animate__fadeIn">
            <input 
              type="text"
              name="search"
              placeholder="Find your hero..."
              autoComplete="off"
              className="form-control"
              value={search}
              onChange={ handleInputChange }
            />
            <div className="d-grid">
              <button
                type="submit"
                className="btn btn-outline-success mt-2"
              >
                Search
              </button>
            </div>
          </form>
        </div>
        <div className="col-12 col-md-7">
          <h4>Results</h4>
          <hr/>
          <div className="animate__animated animate__fadeIn">

            {
              (q === "") &&
                <div className="alert alert-info">
                  Search a hero
                </div>
            }
            {
              (q !== "" && heroesFiltered.length === 0) &&
                <div className="alert alert-danger">
                  There ir no a hero with "{q}"
                </div>
            }
            {
              heroesFiltered.map(hero => (
                <HeroCard key={hero.id} {...hero} />
              ))
            }
          </div>
        </div>
      </div>
    </div>
  )
}
