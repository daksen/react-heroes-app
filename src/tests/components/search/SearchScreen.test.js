import React from 'react';
import { mount } from 'enzyme';
import { SearchScreen } from '../../../components/search/SearchScreen';
import { MemoryRouter, Route } from 'react-router';
import reactRouter from 'react-router';


describe('Pruebas en <SearchScreen />', () => {

  test('debe de mostrarse correctamente con valores por defecto', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path="/search">
          <SearchScreen />
        </Route>
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("div.alert-info").text().trim()).toBe("Search a hero");
  });

  test('debe de mostrar el search y el input con el valor del queryString', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Route path="/search">
          <SearchScreen />
        </Route>
      </MemoryRouter>
    );
    expect(wrapper.find("input").prop("value")).toBe("batman");
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de mostrar un error si no se encuentra el heroe', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=test"]}>
        <Route path="/search">
          <SearchScreen />
        </Route>
      </MemoryRouter>
    );
    expect(wrapper.find("div.alert-danger").text().trim()).toBe('There ir no a hero with "test"');
  });

  test('debe de llamar el push del history', () => {
    const mockPush = jest.fn();
    reactRouter.useHistory = jest.fn().mockReturnValue({
      push: mockPush,
    });
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path="/search">
          <SearchScreen />
        </Route>
      </MemoryRouter>
    );
    wrapper.find("input").simulate("change", {
      target: {
        name: "search",
        value: "batman"
      }
    });
    wrapper.find("form").prop("onSubmit")({
      preventDefault(){}
    });
    expect(mockPush).toHaveBeenCalledWith("?q=batman");
  });
  
});
