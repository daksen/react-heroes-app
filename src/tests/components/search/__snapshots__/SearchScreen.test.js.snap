// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrar el search y el input con el valor del queryString 1`] = `
<div
  className="container mt-5"
>
  <div
    className="row"
  >
    <div
      className="col-12 col-md-5 mb-4 mb-md-0"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        className="animate__animated animate__fadeIn"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero..."
          type="text"
          value="batman"
        />
        <div
          className="d-grid"
        >
          <button
            className="btn btn-outline-success mt-2"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-12 col-md-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="animate__animated animate__fadeIn"
      >
        <div
          className="col"
        >
          <div
            className="card"
          >
            <div
              className="row g-4"
            >
              <div
                className="col-md-4"
              >
                <img
                  alt="Batman"
                  className="card-img-top"
                  src="./assets/heroes/dc-batman.jpg"
                />
              </div>
              <div
                className="col-md-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-title"
                  >
                    Batman
                  </h5>
                  <p
                    className="card-text mb-1"
                  >
                    Bruce Wayne
                  </p>
                  <p
                    className="card-text mb-3"
                  >
                    <small
                      className="text-muted"
                    >
                      Detective Comics #27
                    </small>
                  </p>
                  <a
                    href="/hero/dc-batman"
                    onClick={[Function]}
                  >
                    Leer m√°s...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div
  className="container mt-5"
>
  <div
    className="row"
  >
    <div
      className="col-12 col-md-5 mb-4 mb-md-0"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        className="animate__animated animate__fadeIn"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero..."
          type="text"
          value=""
        />
        <div
          className="d-grid"
        >
          <button
            className="btn btn-outline-success mt-2"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-12 col-md-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="animate__animated animate__fadeIn"
      >
        <div
          className="alert alert-info"
        >
          Search a hero
        </div>
      </div>
    </div>
  </div>
</div>
`;
